File: Исходники/Глава 3. Часть 1/14. Дек/Python/Deque.py
class Node:
    def __init__(self, data=None, next=None, prev=None):
        self.next = next
        self.data = data
        self.prev = prev


class Deque:
    def __init__(self):
        self._end = Node()
        self._top = Node(next=self._end)
        self._size = 0

    def _push_first_element(self, data):
        firs_node = Node(data, self._end, self._top)
        self._top.next = firs_node
        self._end.prev = firs_node

    def push_front(self, data):
        if self.is_empty():
            self._push_first_element(data)
        else:
            next_element = self._top.next
            self._top.next = Node(data, next_element, self._top)
            next_element.prev = self._top.next
        self._size += 1

    def pop_back(self):
        if self.is_empty():
            return -1
        prev_end_element = self._end.prev
        before_prev_end = prev_end_element.prev
        self._end.prev = before_prev_end
        before_prev_end.next = self._end
        self._size -= 1
        return prev_end_element.data

    def push_back(self, data):
        if self.is_empty():
            self._push_first_element(data)
        else:
            prev_element = self._end.prev
            self._end.prev = Node(data, self._end, prev_element)
            prev_element.next = self._end.prev
        self._size += 1

    def pop_front(self):
        if self.is_empty():
            return -1
        prev_top_element = self._top.next
        after_prev_top = prev_top_element.next
        self._top.next = after_prev_top
        after_prev_top.prev = self._top
        self._size -= 1
        return prev_top_element.data

    def is_empty(self):
        if (self._top.next == self._end
                or self._end.prev == self._top):
            return True
        return False

    def size(self):
        return self._size

File: Исходники/Глава 3. Часть 1/14. Дек/C++/Продуктовый/Deque.cpp
#include <iostream>

#include "Deque.h"

using namespace std;

bool Deque::is_empty() {
	if (top == nullptr) return true;
	return false;
}

void Deque::push_back(int element) {
	Box* new_element = new Box(element);
	new_element->next = top;
	top = new_element;
	count += 1;
}

void Deque::push_front(int element) {
	Box* new_element = new Box(element);
	if (is_empty()) {
		new_element->next = top;
		top = new_element;
	} else {
		new_element->data = element;
		Box* curr_element = top;
		while (curr_element->next != end) {
			curr_element = curr_element->next;
		}
		curr_element->next = new_element;
	}
	count += 1;
}

int Deque::pop_back() {
	if (is_empty()) return -1;
	Box* front_element = top;
	top = front_element->next;
	count -= 1;
	return front_element->data;
}

int Deque::pop_front() {
	if (is_empty()) return -1;
	Box* curr_element = top;
	Box* prev_element = new Box();
	while (curr_element->next != end) {
		prev_element = curr_element;
		curr_element = curr_element->next;
	}
	prev_element->next = end;
	if (size() == 1) top = nullptr;
	count -= 1;
	return curr_element->data;
}

int Deque::size() {
	return count;
}